// myfatoorah-toolkit/dist/config/api.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleAxiosError = exports.createApiInstance = void 0;
const axios_1 = __importDefault(require("axios"));
const createApiInstance = (apiUrl, apiTokenKey) => {
    return axios_1.default.create({
        baseURL: apiUrl,
        headers: {
            Accept: 'application/json',
            Authorization: `Bearer ${apiTokenKey}`,
            'Content-Type': 'application/json'
        }
    });
};
exports.createApiInstance = createApiInstance;
const handleAxiosError = (error) => {
    var _a, _b;
    if (error.response || error.request) {
        console.log('%c Error while processing your request from myfatoorah', 'color: ##ff0000');
        return ((_a = error.response) === null || _a === void 0 ? void 0 : _a.data) || ((_b = error.response) === null || _b === void 0 ? void 0 : _b.statusText);
    }
    console.log('%c Error has occurred', error, 'color: ##ff0000');
    process.exit(1);
};
exports.handleAxiosError = handleAxiosError;
